### ISSUE TITLE
Not displaying card content of public board: Snap, Docker and Sandstorm Shared Wekan Board Link

### ISSUE THREAD

**Issue by ertanalytics** [wekan/wekan#1623]
## Issue

**Server Setup Information**:

* Operating System: Ubuntu 16.04
* Deployment Method(snap/sandstorm/mongodb bundle): Sandstorm
* Http frontend (Caddy, Nginx, Apache, see config examples from Wekan GitHub wiki first):Nginx
* Node Version:Latest Sandstorm Wekan release as of 5/3/18
* MongoDB Version:Latest Sandstorm Wekan release as of 5/3/18
* ROOT_URL environment variable http(s)://(subdomain).example.com(/suburl):
https://ertanalytics.us/shared/JXWpB2JhTjG3c2gk5oDscntAzc7CWylxf7blN2P2QqQ
![cards not visible](https://user-images.githubusercontent.com/7130518/39560862-2fc7564a-4e67-11e8-8f5b-7c6ba03b3194.JPG)

**Problem description**:
When providing historical shared read only links to users, the page does not load to display cards.  You can see the activity bar on the right, but cannot select cards that allow them to display details.

- *be as explicit as you can*
- *describe the problem and its symptoms*
- *explain how to reproduce*
- *attach whatever information that can help understanding the context (screen capture, log files in .zip file)*

**Comment by xet7** [comment]
I did see that some versions ago, but not after it.

[Try Wekan version from experimental Sandstorm App Market](https://apps.sandstorm.io/app/m86q05rdvj14yvn78ghaxynqz7u2svw6rnttptxx49g1785cdv1h?experimental=true)

**Comment by xet7** [comment]
Anyway, I've been thinking about enabling seeing all boards at Sandstorm version of Wekan, and removing more of the differences between Standalone and Sandstorm Wekan. Another reason for that is, when importing Wekan board, it says `Board not found`, and when importing new boards, there really is more that one board.

**Comment by xet7** [comment]
There could also be need to have some of the admin panel functionality inside Wekan grain itself, because I don't know will it be possible to remove users from Wekan board from Sandstorm menus itself.

**Comment by ocdtrekkie** [comment]
Yeah, I owe you some app approvals there. I just did my last major homework assignments for my current course tonight, so this weekend is looking good. :)

I am a little concerned about adding more bloat to the interface (given that the Sandstorm interface is also there), but there's not inherently anything wrong with moving in that direction. In an example: One of Sandstorm's Git apps has been modified to only permit one repo, and another allows multiple repos in a grain.

This may take some reworking of permissions though: Sandstorm is going to expect that everyone in a given grain has the permissions Sandstorm is giving it, whereas Wekan boards might have different permissions. This would confuse people, and reduces the security benefit of Sandstorm, since a Wekan bug could let someone with access to one board get to another board in the same grain.

I also don't think having the admin panel is a bad thing, it would just require some rearrangement of where you check for Sandstorm, since some admin panels will never make sense in Sandstorm, you probably want to hide them. (I feel Rocket.Chat can be a bit confusing because it's whole admin panel shows up in Sandstorm, but many of the options do not work.)

**Comment by xet7** [comment]
If I get help to make user management working directly on Sandstorm side, it would sure be better.

**Comment by xet7** [comment]
I could not even get Wekan API enabled on Sandstorm at #1279 , because I could not figure out how to make permissions correct for API. I did just get permission denied errors, tried to add detecting sandstorm headers etc, and did not get it working yet correctly.

 So for that, and also Sandstorm user management, I would need to have someone giving tips to where in code to try to do modifications.

**Comment by xet7** [comment]
@ertanalytics 

Did upgrading Wekan help? There is upgrade popup about Wekan in Sandstorm installs, because it's now in official Sandstorm App Market.

**Comment by hpvb** [comment]
I just tried creating a new board on my sandstorm installation and I'm seeing the same issue. Read-only links to boards are empty with the view menu displaying null.

**Comment by xet7** [comment]
@hpvb 

Please add animated gif video as comment to this issue, about what is happening.

**Comment by hpvb** [comment]
Here you go!

![wekan](https://user-images.githubusercontent.com/2591099/40053598-2f4b5a6a-5842-11e8-8ce9-030c7f240bc3.gif)

**Comment by xet7** [comment]
You did close "Sign In" popup. Sharing is only to those users that can login to same server, and click Sign-in. It is not shared to anonymous users, that feature only works in Standalone Wekan. In Sandstorm Admin, it's also possible to limit for example that only users from example.com domain using Google Auth can login.

![sandstorm-share-reguires-sign-in](https://user-images.githubusercontent.com/15545/40056357-d56b8ecc-5853-11e8-8e40-e74d1c575a51.png)

![sandstorm-share-logged-in-access](https://user-images.githubusercontent.com/15545/40056328-c0b0c6c8-5853-11e8-9ee4-e0004124d5cb.png)

**Comment by hpvb** [comment]
OK, I'd expect a read-only link to work for non-logged in users. It does for other apps. I guess this would be a feature request then? :)

**Comment by xet7** [comment]
I don't know is this bug in Wekan or Sandstorm, as there has been a lot of changes, or has this feature worked sometime before.

**Comment by ocdtrekkie** [comment]
That's gotta pertain to how Wekan handles the new board/list view setting being user-specific. You've got to figure out what Wekan considers an anonymous Sandstorm viewer, and assign them a default view. What does Wekan standalone do when someone views the board without signing in?

**Comment by ocdtrekkie** [comment]
AFAIK, non-signed in users used to be able to see boards from sharing links without signing in. But we'd have to pull a version of the SPK from before the board/list view being profile-specific to test that.

I would consider anywhere something comes back as "null" to be a bug.

**Comment by xet7** [comment]
In Stanalone Wekan there is Public/Private button at board header where board name and search button is, like can be seen at https://wekan.github.io screenshot. I'll try making Private/Public button visible at Sandstorm to see does it help.

**Comment by ocdtrekkie** [comment]
sandstorm.js is supposed to force Wekan to public mode, since the security is enforced by Sandstorm itself.

See here: https://github.com/wekan/wekan/blob/devel/sandstorm.js#L317

**Comment by ocdtrekkie** [comment]
Essentially, it's the same problem that was fixed here: https://github.com/wekan/wekan/commit/c4b898633e1c959f51815ee93a354a093f1139b6 but for Sandstorm anonymous users.

**Comment by ertanalytics** [comment]
This was working prior to the upgrade I did on 5/3/2018 to answer xet7.  The shared link could be provided without the user having to sign in and could stay anonymous so long as they visited the provided link.

**Comment by ocdtrekkie** [comment]
@andresmanelli Was curious if you might have any insight on where we can apply the board view to fix this, since you fixed a similar bug.

**Comment by andresmanelli** [comment]
@ocdtrekkie I'm currently quite busy but I have plans to continue contributing to wekan in the next days/weeks.

**Comment by luxas0072** [comment]
So can i change something in my configuration or must i wait a next update to fix this bug?

**Comment by xet7** [comment]
@luxas0072 

Anyone that has time can fix this bug, see:
https://blog.wekan.team/2018/05/wekan-v1-00-released/index.html

**Comment by luxas0072** [comment]
@xet7 i tried your tutorial here: https://github.com/wekan/wekan/wiki/Install-and-Update#updating-docker
I did it once everything worked but i was still suck to version 0.89.
I tried to do the tutorial a second time and i'm getting this error message when i try to execute this command :
docker rmi wekanteam/wekan:latest
`docker rmi wekanteam/wekan:latest
Error response from daemon: conflict: unable to remove repository reference "wekanteam/wekan:latest" (must force) - container d6bc973ef246 is using its referenced image aaf8819aee47
` [Image here](https://gyazo.com/6fc5b9d3e40d564153cb72a6ffed1137)

the weird thing is that when i try to execute this command:
[docker images](https://gyazo.com/92219421fed832086269c5eb9125c12e)
It says my version of wekan is the latest which isn't the case: https://gyazo.com/d3a4d105ebb4c05a934cdef894fb69fd

**Comment by luxas0072** [comment]
should i start another topic for this ?

**Comment by xet7** [comment]
@luxas0072 

Yes, if you still have problems. Anyway, docker hub image is usually broken, as you see on [releases page](https://github.com/wekan/wekan/releases), so you should use quay image. And, snap is easier to install than Docker, and snap updates automatically.

**Comment by luxas0072** [comment]
@xet7 i'll try that this monday i'm not at work anymore have a good week end and thx for your help :)

**Comment by luxas0072** [comment]
thanks for your help but i installed with snap in version 1.02 : https://gyazo.com/05ffb1e7d1687b0ba82a919a25890492
but i always have this bug :/ https://gyazo.com/a9c51929c6bd4c9e433f61c2aceb0383
here is my configuration https://gyazo.com/2879da273542464d55cfdc9a4528ebd7

**Comment by d3dbit** [comment]
Hello all,
I have the same issue. If I try to open a shared link, the board appears blank. It occurs for both public and private boards, and for a standalone Wekan environment (docker - v1.02).
Thanks.

**Comment by chasgames** [comment]
Same here, not working in docker using quay.io/wekan/wekan latest tag.

Wekan v1.02

**Comment by ChronikEwok** [comment]
Hi all, 
I'm new here so firstly, thanks to everyone who has contributed to this project, it's a great tool.

I was looking into this issue and hopefully the below helps.
From my understanding, when the boardBody.jade template attempts to execute the below lines, there are issues with the isViewSwimlanes and isViewLists functions.

![image](https://user-images.githubusercontent.com/8584218/40848783-1126106a-65b8-11e8-8a2f-6e7922efb423.png)

When you look at those functions in boardBody.js, they both rely on having a currentUser. Obviously in the case of an anonymous user, there is no currentUser value, so both functions will return false, and the template returns neither view. 

![image](https://user-images.githubusercontent.com/8584218/40848841-3d83125c-65b8-11e8-822d-306f0f0f1d7d.png)

There are a few solutions, but since we're viewing the board in either Swimlane OR List view, the template should perhaps be an 'if else' statement as follows:

![image](https://user-images.githubusercontent.com/8584218/40848901-661a0932-65b8-11e8-9d23-4b58c6520f51.png)

Consequently, it will default to List View when you're viewing as an anonymous user. 
Alternatively, the isViewSwimlanes and isViewLists functions should have some sort of exception for when there is no user.

I'm relatively new to Wekan/Meteor, so please review this fix and whether it violates any permission settings and apply as needed.

Also, I haven't fully tested this, but there is at least one other instance where this issue with an undefined currentUser breaks some functionality when a non-user is viewing a public board. That instance being when clicking on a card to view its details. This arises at the start of the swimlane.js component in the currentCardIsInThisList function:

![image](https://user-images.githubusercontent.com/8584218/40849070-defba888-65b8-11e8-9d2f-be4be7058df5.png)

I don't fully understand the reason behind the logic here, as I'm not sure why 'false' is an option to return, but the developer of this component probably has a valid explanation. However to be able to view the card details, you can again rearrange the 'if else' statement and add an exception as follows:

![image](https://user-images.githubusercontent.com/8584218/40849443-def1ccfe-65b9-11e8-82ec-8886b284892e.png)

However, there is still an issue with switching to Swimlane View, as you still get that 'null' text displayed beside the icon as mentioned previously. 

Hopefully this helps, but let me know if it needs further looking into. I'm currently using Wekan on a project so hopefully can contribute more in future.

**Comment by ocdtrekkie** [comment]
@ChronikEwok You indeed have a very solid understanding of the issue. And my impression is Wekan *should* default to list view if a user hasn't selected otherwise. For the first issue, I'd definitely like to see that fix.

For the second one, my guess is "return false" needs to be an option because a function named currentCardIsInThisList should be able to return in the negatory if the current card is in some other list.

**Comment by ChronikEwok** [comment]
Thanks @ocdtrekkie 

With the second one, the 'currentCard.listId === listId' statement should catch whether it's in the correct list or not. That final else statement seems to catch if there is no currentUser.profile.boardView value, which unfortunately means it won't show card details for non-users. The amendment I made will default to List View if there is no user.

In general, when the swimlanes were configured, they didn't account for public boards where there is no user. So as a user, the board view can be toggled by changing the currentUser.profile.boardView value. But in order to toggle the board view as a non-user, perhaps the best solution is to setup a session property that can be toggled?

**Comment by d3dbit** [comment]
Hello all,
With the v1.3 I am experimenting the same issue. If I try to open a shared link, the board appears blank. It occurs for both public and private boards.
Thanks.

**Comment by chasgames** [comment]
Although can now access public board, cannot access content of card

```Exception in template helper: TypeError: Cannot read property 'profile' of null```

**Comment by ChronikEwok** [comment]
With the issue of not being able to display card details, you can look at the second half of my first post here for where the issue is in the code and the fix I used for it. Hope that helps.

**Comment by xet7** [comment]
@ChronikEwok 

Can you make a pull request?

https://help.github.com/articles/fork-a-repo/

https://help.github.com/articles/creating-a-pull-request-from-a-fork/

**Comment by d3dbit** [comment]
Hello all,
I think that with the v1.11.0 I have solved the issue. 
Now I can open a board form a shared link (for example, when I receive an email with a link, if I press it I can access the board).
Thanks for your work.

**Comment by 7adietri** [comment]
We're on Snap 1.11 now and it still looks broken. The swimlanes toggle says `null` and when clicking on cards, they shift a few pixels to the right without opening.

**Comment by chasgames** [comment]
I'm on Docker using Quay.io Wekan 1.11, still seeing a few problems when board set to public
- Swim lanes not showing at all
- When you click on a card nothing happens, I can't see the content of the card.
- I see ```null```, but where your profile name would usually be.

Thanks!

**Comment by xet7** [comment]
@ChronikEwok

Thanks for [your fix](https://github.com/ChronikEwok/wekan/commit/cad9b20451bb6149bfb527a99b5001873b06c3de)! Would you like to have [$15 bounty](https://www.bountysource.com/issues/57883284-not-displaying-card-content-of-public-board-snap-docker-and-sandstorm-shared-wekan-board-link) from this issue? If yes, you can [claim it at BountySource](https://www.bountysource.com/issues/57883284-not-displaying-card-content-of-public-board-snap-docker-and-sandstorm-shared-wekan-board-link).

Another alternative is to give that bounty to me, to support my maintaining of Wekan.

What would you like to do?

**Comment by ocdtrekkie** [comment]
WOW. This one's been out here a long time. Thanks @ChronikEwok !

**Comment by ChronikEwok** [comment]
> 
> 
> @ChronikEwok
> 
> Thanks for [your fix](https://github.com/ChronikEwok/wekan/commit/cad9b20451bb6149bfb527a99b5001873b06c3de)! Would you like to have [$15 bounty](https://www.bountysource.com/issues/57883284-not-displaying-card-content-of-public-board-snap-docker-and-sandstorm-shared-wekan-board-link) from this issue? If yes, you can [claim it at BountySource](https://www.bountysource.com/issues/57883284-not-displaying-card-content-of-public-board-snap-docker-and-sandstorm-shared-wekan-board-link).
> 
> Another alternative is to give that bounty to me, to support my maintaining of Wekan.
> 
> What would you like to do?

Happy for you to have the bounty @xet7 
Thanks for maintaining this great project.
